pkg/reconciler/finalizer/controller.go

```diff
@@ -0,0 +1,168 @@
+func (c *Controller) newExampleObject() client.Object {
+	if c.SyncScope == declared.RootReconciler {
+		exampleObj := &v1beta1.RootSync{}
+		exampleObj.Name = c.SyncName
+		exampleObj.Namespace = configmanagement.ControllerNamespace
+		return exampleObj
+	}
+	exampleObj := &v1beta1.RepoSync{}
+	exampleObj.Name = c.SyncName
+	exampleObj.Namespace = string(c.SyncScope)
+	return exampleObj
+}
+
+// Reconcile responds to changes in the RootSync/RepoSync being watched.
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	// Returned error triggers retry, so we rarely need to populate the result.
+	var result reconcile.Result
+
+	rs := c.newExampleObject()
+	rsKey := client.ObjectKeyFromObject(rs)
+	reqKey := req.NamespacedName
+
+	// This should never happen, if the controller is configured correctly.
+	if reqKey.Name != rsKey.Name || reqKey.Namespace != rsKey.Namespace {
+		klog.Errorf("Controller misconfiguration: reconciler called for the wrong %T: expected %q, but got %q",
+			rs, rsKey, reqKey)
+		// Programmer error. Retry won't help, so don't return the error.
+		return result, nil
+	}
+
+	// Get the latest RSync from the client cache
+	if err := c.Client.Get(ctx, rsKey, rs); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Do nothing. After the RSync is NotFound is too late to finalize.
+			// The finalizer should have run already in OnUpdate when the
+			// deletionTimestamp was added. If we tried to run the destroyer
+			// now, it would probably fail, because the reconciler-manager will
+			// probably have deleted the dependencies. And even if the
+			// dependencies were still around, the reconciler will soon be
+			// deleted by the reconciler-manager, if it hasn't already.
+			return result, nil
+		}
+		return result, status.APIServerError(err, fmt.Sprintf("failed to get %s", objSummary(rs)))
+	}
+
+	if !rs.GetDeletionTimestamp().IsZero() {
+		// Object being deleted.
+		if controllerutil.ContainsFinalizer(rs, metadata.ReconcilerFinalizer) {
+			if err := c.Finalizer.Finalize(ctx, rs); err != nil {
+				return result, errors.Wrapf(err, "finalizing")
+			}
+		}
+	} else {
+		if err := c.reconcileFinalizer(ctx, rs); err != nil {
+			return result, errors.Wrapf(err, "reconciling finalizer")
+		}
+	}
+	return result, nil
+}
+
+// reconcileFinalizer adds or removes the `configsync.gke.io/reconciler`
+// finalizer, depending on the existance and value of the
+// `configsync.gke.io/deletion-propagation-policy` annotation.
+func (c *Controller) reconcileFinalizer(ctx context.Context, obj client.Object) error {
+	policyStr, found := obj.GetAnnotations()[metadata.DeletionPropagationPolicyAnnotationKey]
+	policy := metadata.DeletionPropagationPolicy(policyStr)
+	if !found {
+		// Orphan is the default policy
+		policy = metadata.DeletionPropagationPolicyOrphan
+	}
+	switch policy {
+	case metadata.DeletionPropagationPolicyForeground:
```

Comment:
Do you expect we will just support the binary on/off finalizer, or might we support different types of finalizers in the future (e.g. background deletion)? I'm asking because I'm wondering how well this model would fit with multiple types of finalizers

Reply:
I designed it for use with background as an option, but it turns out we don't have a use case for it, so i removed it. That was from a previous thread with Nan.

As for multiple finalizers, the other reconciler-manager finalizer (https://github.com/GoogleContainerTools/kpt-config-sync/pull/17) doesn't affect user objects, only the reconciler deployment and its dependencies, so it won't have an on-off switch or any config. It will just always be foreground, instead of background (current impl).